src/util/unesc.js

```diff
@@ -1,19 +1,96 @@
+    "@", "[", "]", "^", "`", "{", "|", "}", "~", "_", ":",
+]);
+
+const CSS_SPECIAL_CHARS_HEX_LONG = {}; 
+const CSS_SPECIAL_CHARS_HEX_SHORT = {};
+
+for (const char of Array.from(CSS_SPECIAL_CHARS)) {
+    CSS_SPECIAL_CHARS_HEX_LONG[`0000${char.charCodeAt(0).toString(16)}`] = char;
+    CSS_SPECIAL_CHARS_HEX_SHORT[`${char.charCodeAt(0).toString(16)}`] = char;
+}
+
+function matchesOctal (s) {
+    const short = s[0] + s[1];
+    if (CSS_SPECIAL_CHARS_HEX_SHORT[short]) {
+        return {
+            char: CSS_SPECIAL_CHARS_HEX_SHORT[short],
+            length: 3,
+        };
+    } 
+    const long = s[0] + s[1] + s[2] + s[3] + s[4] + s[5];
+    if (CSS_SPECIAL_CHARS_HEX_LONG[long]) {
+        return {
+            char: CSS_SPECIAL_CHARS_HEX_LONG[long],
+            length: 6,
+        };
+    }
+
+    return undefined;
+}
 
 export default function unesc (str) {
-    return str.replace(unescapeRegExp, (_, escaped, escapedWhitespace) => {
-        const high = '0x' + escaped - 0x10000;
-
-        // NaN means non-codepoint
-        // Workaround erroneous numeric interpretation of +"0x"
-        // eslint-disable-next-line no-self-compare
-        return high !== high || escapedWhitespace
-            ? escaped
-            : high < 0
-                ? // BMP codepoint
-                String.fromCharCode(high + 0x10000)
-                : // Supplemental Plane codepoint (surrogate pair)
-                String.fromCharCode((high >> 10) | 0xd800, (high & 0x3ff) | 0xdc00);
-    });
+    let ret = "";
+
+    for (let i = 0; i < str.length; i++) {
+        if (str[i] === "\\" && !isNaN(Number(str[i + 1]))) {
+            // Handle the \3 leading digit escape case.
+            if (str[i + 1] === "3" && Number(str[i + 2]) <= 9) {
+                ret += str[i + 2];
+                i += 3;
+                continue;
+            // Special case the ":" handling .. thus the "A" check.
+            } else if (str[i +2] === "A") {
+                ret += ":";
+                i += 3;
+                continue;
+            } else {
+                const match = matchesOctal(str.slice(i+1, i + 7));
+                if (match) {
+                    ret += match.char;
+                    i += match.length;
+                    continue;
+                }
+            } 
+        }
+
+        if ((str[i] === "\\")) {
+            // if // is at the end of the string retain it
+            // https://github.com/postcss/postcss-selector-parser/commit/01a6b346e3612ce1ab20219acc26abdc259ccefb
+            if (str.length === i + 1) {
+                ret += str[i];
+            }
+
+            // Retain a pair of \\ if double escaped `\\\\`
+            // https://github.com/postcss/postcss-selector-parser/commit/268c9a7656fb53f543dc620aa5b73a30ec3ff20e
+            if (str[i +1] === "\\") {
+                ret += "\\";
+                i++;
+                continue;
+            }
+
+            // Need to check if we have an emoji
```

Comment:
I might be misunderstanding, but I think the intention here is to detect astral symbols (i.e. code points above U+FFFF, i.e. those in the range from U+010000 to U+10FFFF). This is orthogonal to the use of emoji. (Some emoji have a code point `<` U+FFFF.) Did I get that right?

In that case, there could be up to 6 hex digits (and that's [the maximum](https://drafts.csswg.org/css-syntax/#consume-escaped-code-point) that CSS allows as well).